Mathlib/ModelTheory/Equivalence.lean

```diff
@@ -34,31 +33,129 @@ variable {M : Type*} [Nonempty M] [L.Structure M] [M ⊨ T]
 
 namespace Theory
 
+/-- `φ ⟹[T] ψ` indicates that `φ` implies `ψ` in models of `T`. -/
+protected def Imp (T : L.Theory) (φ ψ : L.BoundedFormula α n) : Prop :=
+  T ⊨ᵇ φ.imp ψ
+
+@[inherit_doc FirstOrder.Language.Theory.Imp]
+scoped[FirstOrder] notation:51 φ:50 " ⟹[" T "] " ψ:51 => Language.Theory.Imp T φ ψ
+
+namespace Imp
+
+@[refl]
+protected theorem refl (φ : L.BoundedFormula α n) : φ ⟹[T] φ := fun _ _ _ => id
+
+instance : IsRefl (L.BoundedFormula α n) T.Imp := ⟨Imp.refl⟩
+
+@[trans]
+protected theorem trans {φ ψ θ : L.BoundedFormula α n} (h1 : φ ⟹[T] ψ) (h2 : ψ ⟹[T] θ) :
+    φ ⟹[T] θ := fun M v xs => (h2 M v xs) ∘ (h1 M v xs)
+
+instance : IsTrans (L.BoundedFormula α n) T.Imp := ⟨fun _ _ _ => Imp.trans⟩
+
+end Imp
+
+section Imp
+
+lemma bot_imp (φ : L.BoundedFormula α n) : ⊥ ⟹[T] φ := fun M v xs => by
+  simp only [BoundedFormula.realize_imp, BoundedFormula.realize_bot, false_implies]
+
+lemma imp_top (φ : L.BoundedFormula α n) : φ ⟹[T] ⊤ := fun M v xs => by
+  simp only [BoundedFormula.realize_imp, BoundedFormula.realize_top, implies_true]
+
+lemma imp_sup_left (φ ψ : L.BoundedFormula α n) : φ ⟹[T] φ ⊔ ψ := fun M v xs => by
```

Comment:
I'd prefer to name this `sup_imp_left` in analogy with `sup_le_left` for orders

Reply:
(and similar below)

Reply:
You mean [`le_sup_left](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Order/Lattice.html#le_sup_left)`?

Reply:
oops lol